File: libs/trading_models/patterns.py
"""Pattern recognition models."""

from datetime import datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import Field, field_validator

from .base import BaseModel
from .enums import PatternType, Timeframe


class PatternHit(BaseModel):
    """Detected pattern with confidence and metadata."""

    pattern_id: str = Field(..., description="Unique pattern identifier")
    pattern_type: PatternType = Field(..., description="Type of pattern detected")
    symbol: str = Field(..., description="Trading symbol")
    timeframe: Timeframe = Field(..., description="Timeframe where pattern was detected")
    timestamp: datetime = Field(..., description="Pattern detection timestamp")

    # Pattern confidence and scoring
    confidence: float = Field(..., ge=0, le=1, description="Pattern confidence (0-1)")
    strength: float = Field(..., ge=0, le=10, description="Pattern strength (0-10)")

    # Price levels
    entry_price: Optional[Decimal] = Field(None, gt=0, description="Suggested entry price")
    stop_loss: Optional[Decimal] = Field(None, gt=0, description="Suggested stop loss")
    take_profit: Optional[Decimal] = Field(None, gt=0, description="Suggested take profit")

    # Support/Resistance levels
    support_levels: list[Decimal] = Field(default_factory=list, description="Support price levels")
    resistance_levels: list[Decimal] = Field(default_factory=list, description="Resistance price levels")

    # Pattern-specific data
    pattern_data: dict[str, Any] = Field(default_factory=dict, description="Pattern-specific metadata")

    # Validation context
    bars_analyzed: int = Field(..., gt=0, description="Number of bars used in analysis")
    lookback_period: int = Field(..., gt=0, description="Lookback period for pattern detection")

    # Performance tracking
    historical_win_rate: Optional[float] = Field(None, ge=0, le=1, description="Historical win rate for this pattern")
    avg_return: Optional[float] = Field(None, description="Average return for this pattern")

    @field_validator("support_levels", "resistance_levels")
    @classmethod
    def validate_price_levels(cls, v):
        """Validate price levels are positive and sorted."""
        if v:
            if any(level <= 0 for level in v):
                raise ValueError("All price levels must be positive")
            if v != sorted(v):
                raise ValueError("Price levels must be sorted")
        return v

    @field_validator("stop_loss")
    @classmethod
    def validate_stop_loss(cls, v, info):
        """Validate stop loss relative to entry price."""
        if v and info.data and "entry_price" in info.data and info.data["entry_price"]:
            entry = info.data["entry_price"]
            # For long positions, stop should be below entry
            # For short positions, stop should be above entry
            # We'll validate this is reasonable (within 50% of entry)
            diff_pct = abs(float(v - entry) / float(entry))
            if diff_pct > 0.5:
                raise ValueError("Stop loss too far from entry price (>50%)")
        return v

    @field_validator("take_profit")
    @classmethod
    def validate_take_profit(cls, v, info):
        """Validate take profit relative to entry price."""
        if v and info.data and "entry_price" in info.data and info.data["entry_price"]:
            entry = info.data["entry_price"]
            # Validate reasonable take profit (within 200% of entry)
            diff_pct = abs(float(v - entry) / float(entry))
            if diff_pct > 2.0:
                raise ValueError("Take profit too far from entry price (>200%)")
        return v


class PatternCollection(BaseModel):
    """Collection of patterns for a symbol/timeframe."""

    symbol: str = Field(..., description="Trading symbol")
    timeframe: Timeframe = Field(..., description="Analysis timeframe")
    timestamp: datetime = Field(..., description="Analysis timestamp")

    patterns: list[PatternHit] = Field(default_factory=list, description="Detected patterns")

    # Summary statistics
    total_patterns: int = Field(0, ge=0, description="Total number of patterns")
    avg_confidence: float = Field(0, ge=0, le=1, description="Average pattern confidence")
    strongest_pattern: Optional[str] = Field(None, description="ID of strongest pattern")

    def add_pattern(self, pattern: PatternHit) -> None:
        """Add a pattern to the collection."""
        self.patterns.append(pattern)
        self._update_stats()

    def _update_stats(self) -> None:
        """Update summary statistics."""
        self.total_patterns = len(self.patterns)
        if self.patterns:
            self.avg_confidence = sum(p.confidence for p in self.patterns) / len(self.patterns)
            strongest = max(self.patterns, key=lambda p: p.strength)
            self.strongest_pattern = strongest.pattern_id
        else:
            self.avg_confidence = 0.0
            self.strongest_pattern = None

    def get_patterns_by_type(self, pattern_type: PatternType) -> list[PatternHit]:
        """Get patterns of a specific type."""
        return [p for p in self.patterns if p.pattern_type == pattern_type]

    def get_high_confidence_patterns(self, min_confidence: float = 0.7) -> list[PatternHit]:
        """Get patterns above confidence threshold."""
        return [p for p in self.patterns if p.confidence >= min_confidence]

File: libs/rust-common/src/trading_models/patterns.rs
//! Pattern recognition structures.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::enums::{PatternType, Timeframe};

/// Detected pattern with confidence and metadata.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatternHit {
    pub pattern_id: String,
    pub pattern_type: PatternType,
    pub symbol: String,
    pub timeframe: Timeframe,
    pub timestamp: DateTime<Utc>,
    
    // Pattern confidence and scoring
    pub confidence: f64,
    pub strength: f64,
    
    // Price levels
    pub entry_price: Option<f64>,
    pub stop_loss: Option<f64>,
    pub take_profit: Option<f64>,
    
    // Support/Resistance levels
    pub support_levels: Vec<f64>,
    pub resistance_levels: Vec<f64>,
    
    // Pattern-specific data
    pub pattern_data: HashMap<String, serde_json::Value>,
    
    // Validation context
    pub bars_analyzed: u32,
    pub lookback_period: u32,
    
    // Performance tracking
    pub historical_win_rate: Option<f64>,
    pub avg_return: Option<f64>,
}

impl PatternHit {
    /// Validate pattern data.
    pub fn validate(&self) -> Result<(), String> {
        // Validate confidence range
        if !(0.0..=1.0).contains(&self.confidence) {
            return Err("Confidence must be between 0 and 1".to_string());
        }
        
        // Validate strength range
        if !(0.0..=10.0).contains(&self.strength) {
            return Err("Strength must be between 0 and 10".to_string());
        }
        
        // Validate positive prices
        if let Some(entry) = self.entry_price {
            if entry <= 0.0 {
                return Err("Entry price must be positive".to_string());
            }
        }
        
        if let Some(stop) = self.stop_loss {
            if stop <= 0.0 {
                return Err("Stop loss must be positive".to_string());
            }
        }
        
        if let Some(target) = self.take_profit {
            if target <= 0.0 {
                return Err("Take profit must be positive".to_string());
            }
        }
        
        // Validate price levels are positive and sorted
        for level in &self.support_levels {
            if *level <= 0.0 {
                return Err("Support levels must be positive".to_string());
            }
        }
        
        for level in &self.resistance_levels {
            if *level <= 0.0 {
                return Err("Resistance levels must be positive".to_string());
            }
        }
        
        // Check if levels are sorted
        let mut sorted_support = self.support_levels.clone();
        sorted_support.sort_by(|a, b| a.partial_cmp(b).unwrap());
        if sorted_support != self.support_levels {
            return Err("Support levels must be sorted".to_string());
        }
        
        let mut sorted_resistance = self.resistance_levels.clone();
        sorted_resistance.sort_by(|a, b| a.partial_cmp(b).unwrap());
        if sorted_resistance != self.resistance_levels {
            return Err("Resistance levels must be sorted".to_string());
        }
        
        // Validate historical win rate
        if let Some(win_rate) = self.historical_win_rate {
            if !(0.0..=1.0).contains(&win_rate) {
                return Err("Historical win rate must be between 0 and 1".to_string());
            }
        }
        
        // Validate analysis parameters
        if self.bars_analyzed == 0 {
            return Err("Bars analyzed must be greater than 0".to_string());
        }
        
        if self.lookback_period == 0 {
            return Err("Lookback period must be greater than 0".to_string());
        }
        
        Ok(())
    }
}

/// Collection of patterns for a symbol/timeframe.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatternCollection {
    pub symbol: String,
    pub timeframe: Timeframe,
    pub timestamp: DateTime<Utc>,
    
    pub patterns: Vec<PatternHit>,
    
    // Summary statistics
    pub total_patterns: u32,
    pub avg_confidence: f64,
    pub strongest_pattern: Option<String>,
}

impl PatternCollection {
    /// Create a new pattern collection.
    pub fn new(symbol: String, timeframe: Timeframe) -> Self {
        Self {
            symbol,
            timeframe,
            timestamp: Utc::now(),
            patterns: Vec::new(),
            total_patterns: 0,
            avg_confidence: 0.0,
            strongest_pattern: None,
        }
    }
    
    /// Add a pattern to the collection.
    pub fn add_pattern(&mut self, pattern: PatternHit) {
        self.patterns.push(pattern);
        self.update_stats();
    }
    
    /// Update summary statistics.
    fn update_stats(&mut self) {
        self.total_patterns = self.patterns.len() as u32;
        
        if !self.patterns.is_empty() {
            self.avg_confidence = self.patterns.iter()
                .map(|p| p.confidence)
                .sum::<f64>() / self.patterns.len() as f64;
            
            if let Some(strongest) = self.patterns.iter()
                .max_by(|a, b| a.strength.partial_cmp(&b.strength).unwrap()) {
                self.strongest_pattern = Some(strongest.pattern_id.clone());
            }
        } else {
            self.avg_confidence = 0.0;
            self.strongest_pattern = None;
        }
    }
    
    /// Get patterns of a specific type.
    pub fn get_patterns_by_type(&self, pattern_type: PatternType) -> Vec<&PatternHit> {
        self.patterns.iter()
            .filter(|p| p.pattern_type == pattern_type)
            .collect()
    }
    
    /// Get patterns above confidence threshold.
    pub fn get_high_confidence_patterns(&self, min_confidence: f64) -> Vec<&PatternHit> {
        self.patterns.iter()
            .filter(|p| p.confidence >= min_confidence)
            .collect()
    }
}